File: express/src/routes/Content/articles.ts
import { Router } from 'express';
import { prisma } from '../../lib/prisma.js';
import { ArticleService } from '../../services/Content/article_service.js';

const router = Router();
const service = new ArticleService();

function validate(data: any): void {
  if (!((data.viewCount == null || data.viewCount >= 0))) throw new Error(`Article view count must not be negative`);
  if (!((data.likesCount == null || data.likesCount >= 0))) throw new Error(`Article likes count must not be negative`);
  if ((data.status === 'PUBLISHED') && !((data.publishedAt === undefined || data.publishedAt != null))) throw new Error(`Published article must have a published_at timestamp`);
}
function applyProjection(obj: any): any {
  if (!obj) return obj;
  const r = { ...obj };
  if ('createdAt' in r) { r.createdAt = r.createdAt; delete r.createdAt; }
  if ('updatedAt' in r) { r.updatedAt = r.updatedAt; delete r.updatedAt; }
  if ('publishedAt' in r) { r.publishedAt = r.publishedAt; delete r.publishedAt; }
  return r;
}

const ALLOWED_TRANSITIONS: Record<string, string[]> = {
  'Draft': ['Published'],
  'Published': ['Archived'],
  'Archived': ['Draft']
};

function assertTransition(current: string, to: string): void {
  const allowed = ALLOWED_TRANSITIONS[current] ?? [];
  if (!allowed.includes(to)) throw new Error(`Transition ${current} -> ${to} is not allowed`);
}

class ArticleLifecycleService {

  async transitionDraftToPublished(id: number): Promise<any> {
    const entity = await prisma.article.findUnique({ where: { id } });
    if (!entity) throw new Error('Article not found: ' + id);
    assertTransition((entity as any).status, 'Published');
    if ((entity as any).title == null) throw new Error('title is required for Draft -> Published');
    if ((entity as any).body == null) throw new Error('body is required for Draft -> Published');
    const updated = await prisma.article.update({ where: { id }, data: { status: 'PUBLISHED' as any } });
    // TODO: entity.publish(); // @after
    return updated;
  }

  async transitionPublishedToArchived(id: number): Promise<any> {
    const entity = await prisma.article.findUnique({ where: { id } });
    if (!entity) throw new Error('Article not found: ' + id);
    assertTransition((entity as any).status, 'Archived');
    const updated = await prisma.article.update({ where: { id }, data: { status: 'ARCHIVED' as any } });
    // TODO: entity.archive(); // @after
    return updated;
  }

  async transitionArchivedToDraft(id: number): Promise<any> {
    const entity = await prisma.article.findUnique({ where: { id } });
    if (!entity) throw new Error('Article not found: ' + id);
    assertTransition((entity as any).status, 'Draft');
    const updated = await prisma.article.update({ where: { id }, data: { status: 'DRAFT' as any } });
    return updated;
  }

  async transitionPublishedToDraft(id: number): Promise<any> {
    const entity = await prisma.article.findUnique({ where: { id } });
    if (!entity) throw new Error('Article not found: ' + id);
    throw new Error('Transition Published -> Draft is not allowed');
  }
}
const lifecycleService = new ArticleLifecycleService();


router.get('/', async (req, res) => {
  const q = req.query.q as string | undefined;
  const where = q ? { OR: [{ title: { contains: q } }, { excerpt: { contains: q } }] } : undefined;
  const items = await prisma.article.findMany(where ? { where } : undefined);
  res.json(items.map(applyProjection));
});

router.post('/', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.title !== undefined) data.title = body.title;
    if (body.slug !== undefined) data.slug = body.slug;
    if (body.body !== undefined) data.body = body.body;
    if (body.excerpt !== undefined) data.excerpt = body.excerpt;
    if (body.coverImageUrl !== undefined) data.coverImageUrl = body.coverImageUrl;
    if (body.status !== undefined) data.status = body.status;
    if (body.articleType !== undefined) data.articleType = body.articleType;
    if (body.language !== undefined) data.language = body.language;
    if (body.viewCount !== undefined) data.viewCount = body.viewCount;
    if (body.likesCount !== undefined) data.likesCount = body.likesCount;
    if (body.isFeatured !== undefined) data.isFeatured = body.isFeatured;
    if (body.publishedAt !== undefined) data.publishedAt = body.publishedAt != null ? new Date(body.publishedAt) : null;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.updatedAt !== undefined) data.updatedAt = body.updatedAt != null ? new Date(body.updatedAt) : null;
    if (body.authorId !== undefined) data.authorId = body.authorId;
    if (body.featuredDeckId !== undefined) data.featuredDeckId = body.featuredDeckId;
  try {
  validate(data);
    const entity = await prisma.article.create({ data });
    res.status(201).json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Validation error') });
  }
});

router.get('/:id', async (req, res) => {
  const entity = await prisma.article.findUnique({ where: { id: Number(req.params.id) } });
  if (!entity) return res.status(404).json({ error: 'Not found' });
  res.json(applyProjection(entity));
});

router.put('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.title !== undefined) data.title = body.title;
    if (body.slug !== undefined) data.slug = body.slug;
    if (body.body !== undefined) data.body = body.body;
    if (body.excerpt !== undefined) data.excerpt = body.excerpt;
    if (body.coverImageUrl !== undefined) data.coverImageUrl = body.coverImageUrl;
    if (body.status !== undefined) data.status = body.status;
    if (body.articleType !== undefined) data.articleType = body.articleType;
    if (body.language !== undefined) data.language = body.language;
    if (body.viewCount !== undefined) data.viewCount = body.viewCount;
    if (body.likesCount !== undefined) data.likesCount = body.likesCount;
    if (body.isFeatured !== undefined) data.isFeatured = body.isFeatured;
    if (body.publishedAt !== undefined) data.publishedAt = body.publishedAt != null ? new Date(body.publishedAt) : null;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.updatedAt !== undefined) data.updatedAt = body.updatedAt != null ? new Date(body.updatedAt) : null;
    if (body.authorId !== undefined) data.authorId = body.authorId;
    if (body.featuredDeckId !== undefined) data.featuredDeckId = body.featuredDeckId;
  try {
  validate(data);
    const entity = await prisma.article.update({ where: { id: Number(req.params.id) }, data });
    res.json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Error') });
  }
});

router.patch('/:id', async (req, res) => {
  const body = req.body;
  const data: any = {};
    if (body.title !== undefined) data.title = body.title;
    if (body.slug !== undefined) data.slug = body.slug;
    if (body.body !== undefined) data.body = body.body;
    if (body.excerpt !== undefined) data.excerpt = body.excerpt;
    if (body.coverImageUrl !== undefined) data.coverImageUrl = body.coverImageUrl;
    if (body.status !== undefined) data.status = body.status;
    if (body.articleType !== undefined) data.articleType = body.articleType;
    if (body.language !== undefined) data.language = body.language;
    if (body.viewCount !== undefined) data.viewCount = body.viewCount;
    if (body.likesCount !== undefined) data.likesCount = body.likesCount;
    if (body.isFeatured !== undefined) data.isFeatured = body.isFeatured;
    if (body.publishedAt !== undefined) data.publishedAt = body.publishedAt != null ? new Date(body.publishedAt) : null;
    if (body.createdAt !== undefined) data.createdAt = body.createdAt != null ? new Date(body.createdAt) : null;
    if (body.updatedAt !== undefined) data.updatedAt = body.updatedAt != null ? new Date(body.updatedAt) : null;
    if (body.authorId !== undefined) data.authorId = body.authorId;
    if (body.featuredDeckId !== undefined) data.featuredDeckId = body.featuredDeckId;
  try {
  validate(data);
    const entity = await prisma.article.update({ where: { id: Number(req.params.id) }, data });
    res.json(applyProjection(entity));
  } catch (err: any) {
    const status = err?.code === 'P2025' ? 404 : err?.code === 'P2002' ? 422 : 400;
    res.status(status).json({ error: err?.code === 'P2002' ? 'Value must be unique' : (err?.message ?? 'Error') });
  }
});

router.post('/:id/publish', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.publish(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/archive', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.archive(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/view', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.increment_view(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.post('/:id/like', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.like(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.delete('/:id/like', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    await service.unlike(id);
    res.status(204).send();
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.get('/:id/reading-time', async (req, res) => {
  const id = Number((req.params as any).id);
  try {
    const result = await service.reading_time_minutes(id);
    res.json({ result });
  } catch (err: any) {
    const status = err?.message?.startsWith('Guard') ? 422 : 404;
    res.status(status).json({ error: err?.message ?? 'Not found' });
  }
});

router.patch('/:id/transitions/draft-to-published', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionDraftToPublished(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/published-to-archived', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionPublishedToArchived(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/archived-to-draft', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionArchivedToDraft(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});

router.patch('/:id/transitions/published-to-draft', async (req, res) => {
  const id = Number(req.params.id);
  try {
    const entity = await lifecycleService.transitionPublishedToDraft(id);
    res.json(entity);
  } catch (err: any) {
    const status = err?.message?.includes('not allowed') || err?.message?.includes('is not allowed') ? 409
      : err?.message?.includes('required') || err?.message?.includes('must be') ? 422
      : 404;
    res.status(status).json({ error: err?.message ?? 'Error' });
  }
});
export default router;

File: gin/internal/handler/content/articles.go
package handler_content

import (
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"

	model "cards_project/internal/model/content"
	"cards_project/internal/handler"
)

type ArticleHandler struct { db *gorm.DB }

func NewArticleHandler(db *gorm.DB) *ArticleHandler {
	return &ArticleHandler{db: db}
}

func (h *ArticleHandler) RegisterRoutes(r gin.IRouter) {
	g := r.Group("/api/articles")
	g.GET("", h.List)
	g.POST("", h.Create)
	g.GET("/:id", h.Get)
	g.PUT("/:id", h.Update)
	g.PATCH("/:id", h.Patch)
	g.POST("/:id/publish", h.Publish)
	g.POST("/:id/archive", h.Archive)
	g.POST("/:id/view", h.IncrementView)
	g.POST("/:id/like", h.Like)
	g.DELETE("/:id/like", h.Unlike)
	g.GET("/:id/reading-time", h.ReadingTimeMinutes)
	g.PATCH("/:id/transitions/draft-to-published", h.TransitionDraftToPublished)
	g.PATCH("/:id/transitions/published-to-archived", h.TransitionPublishedToArchived)
	g.PATCH("/:id/transitions/archived-to-draft", h.TransitionArchivedToDraft)
	g.PATCH("/:id/transitions/published-to-draft", h.TransitionPublishedToDraft)
}

func (h *ArticleHandler) List(c *gin.Context) {
	skip, limit := handler.Paginate(c)
	var rows []model.Article
	q := c.Query("q")
	db := h.db
	if q != "" {
		db = db.Or("title LIKE ?", "%"+q+"%")
		db = db.Or("excerpt LIKE ?", "%"+q+"%")
	}
	if err := db.Offset(skip).Limit(limit).Find(&rows).Error; err != nil {
		handler.DbError(c, err); return
	}
	out := make([]model.ArticleResponse, len(rows))
	for i := range rows { out[i] = rows[i].ToResponse() }
	c.JSON(http.StatusOK, out)
}

func (h *ArticleHandler) Create(c *gin.Context) {
	var req model.ArticleCreateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	if msgs := validateArticle(&req); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	row := model.Article{}
	row.Title = req.Title
	row.Slug = req.Slug
	row.Body = req.Body
	row.Excerpt = req.Excerpt
	row.CoverImageUrl = req.CoverImageUrl
	row.Status = req.Status
	row.ArticleType = req.ArticleType
	row.Language = req.Language
	row.ViewCount = req.ViewCount
	row.LikesCount = req.LikesCount
	row.IsFeatured = req.IsFeatured
	row.PublishedAt = req.PublishedAt
	row.AuthorID = req.AuthorID
	row.FeaturedDeckID = req.FeaturedDeckID
	if err := h.db.Create(&row).Error; err != nil {
		if handler.IsUniqueViolation(err) { handler.UnprocessableError(c, "Value must be unique"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusCreated, row.ToResponse())
}

func (h *ArticleHandler) Get(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *ArticleHandler) Update(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	var req model.ArticleUpdateRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handler.ValidationError(c, err.Error()); return
	}
	row.ApplyUpdate(req)
	createReq := toCreateRequestArticle(&row)
	if msgs := validateArticle(&createReq); len(msgs) > 0 {
		c.JSON(http.StatusBadRequest, gin.H{"errors": msgs}); return
	}
	if err := h.db.Save(&row).Error; err != nil {
		if handler.IsUniqueViolation(err) { handler.UnprocessableError(c, "Value must be unique"); return }
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *ArticleHandler) Patch(c *gin.Context) { h.Update(c) }

func (h *ArticleHandler) Publish(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	err := row.Publish()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *ArticleHandler) Archive(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	err := row.Archive()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *ArticleHandler) IncrementView(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	err := row.IncrementView()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *ArticleHandler) Like(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	err := row.Like()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *ArticleHandler) Unlike(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	err := row.Unlike()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.Status(http.StatusNoContent)
}

func (h *ArticleHandler) ReadingTimeMinutes(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	result, err := row.ReadingTimeMinutes()
	if err != nil { handler.DbError(c, err); return }
	h.db.Save(&row)
	c.JSON(http.StatusOK, gin.H{"result": result})
}

func (h *ArticleHandler) TransitionDraftToPublished(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	if err := row.AssertTransition("Published"); err != nil {
		handler.ConflictError(c, err.Error()); return
	}
	if row.Title == "" {
		handler.UnprocessableError(c, "title is required for this transition"); return
	}
	if row.Body == "" {
		handler.UnprocessableError(c, "body is required for this transition"); return
	}
	row.Status = model.ArticleStatusType_Published
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	_ = row.Publish() // @after
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *ArticleHandler) TransitionPublishedToArchived(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	if err := row.AssertTransition("Archived"); err != nil {
		handler.ConflictError(c, err.Error()); return
	}
	row.Status = model.ArticleStatusType_Archived
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	_ = row.Archive() // @after
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *ArticleHandler) TransitionArchivedToDraft(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	if err := row.AssertTransition("Draft"); err != nil {
		handler.ConflictError(c, err.Error()); return
	}
	row.Status = model.ArticleStatusType_Draft
	if err := h.db.Save(&row).Error; err != nil {
		handler.DbError(c, err); return
	}
	c.JSON(http.StatusOK, row.ToResponse())
}

func (h *ArticleHandler) TransitionPublishedToDraft(c *gin.Context) {
	id, ok := handler.ParseID(c); if !ok { return }
	var row model.Article
	if err := h.db.First(&row, id).Error; err != nil {
		if handler.IsRecordNotFound(err) { handler.NotFound(c, "Article"); return }
		handler.DbError(c, err); return
	}
	handler.ConflictError(c, "transition Published -> Draft is not allowed")
	return
}

// ── Lifecycle hooks ──────────────────────────────────────────────────
func (h *ArticleHandler) hookUpdateSearchIndex(row *model.Article) {
	// TODO: implement update_search_index
}

// ── Validation rules ─────────────────────────────────────────────
func validateArticle(req *model.ArticleCreateRequest) []string {
	var errs []string
	if !((!( req.Status == model.ArticleStatusType_Published ) || (req.PublishedAt != nil))) {
		errs = append(errs, "Published article must have a published_at timestamp")
	}
	if !(req.ViewCount >= 0) {
		errs = append(errs, "Article view count must not be negative")
	}
	if !(req.LikesCount >= 0) {
		errs = append(errs, "Article likes count must not be negative")
	}
	return errs
}

func toCreateRequestArticle(m *model.Article) model.ArticleCreateRequest {
	return model.ArticleCreateRequest{
		Title: m.Title,
		Slug: m.Slug,
		Body: m.Body,
		Excerpt: m.Excerpt,
		CoverImageUrl: m.CoverImageUrl,
		Status: m.Status,
		ArticleType: m.ArticleType,
		Language: m.Language,
		ViewCount: m.ViewCount,
		LikesCount: m.LikesCount,
		IsFeatured: m.IsFeatured,
		PublishedAt: m.PublishedAt,
		AuthorID: m.AuthorID,
		FeaturedDeckID: m.FeaturedDeckID,
	}
}
